src/main/java/net/jitl/common/entity/projectile/base/JEffectCloudEntity.java

```diff
@@ -0,0 +1,259 @@
+            for(float current = 0; current < loops; ++current) {
+               float rotation = this.random.nextFloat() * ((float) Math.PI * 2F);
+               float distance = this.random.nextFloat() * radius; //vanilla has a sqrt here but it seems pointless
+               float xOffset = MathHelper.cos(rotation) * distance;
+               float zOffset = MathHelper.sin(rotation) * distance;
+               int color = getEntityData().get(COLOR);
+               this.level.addAlwaysVisibleParticle(ParticleTypes.ENTITY_EFFECT, this.getX() + xOffset, this.getY(), this.getZ() + zOffset, (color >> 16 & 255) / 255.0, (color >> 8 & 255) / 255.0, (color & 255) / 255.0);
+            }
+		} else { //Server
+			//adjust the radius or delete the entity
+			ImmutablePair<Integer, Float> previousPair = null; //the last specified time-size value the cloud was at
+			ImmutablePair<Integer, Float> nextPair = null; //the time-size value the cloud is moving towards
+			for (int pairCount = 0; pairCount < sizes.size(); pairCount++) {
+				if (sizes.get(pairCount).getKey() < this.tickCount) { //check if the cloud has already been at the specified time-size pair
+					if (pairCount == sizes.size() - 1) { //make sure there is actually another pair (otherwise the cloud has completed its life cycle)
+						this.remove();
+						return;
+					} else if (sizes.get(pairCount + 1).getKey() > this.tickCount) { //the cloud is between this pair and the previous one
+						previousPair = sizes.get(pairCount);
+						nextPair = sizes.get(pairCount + 1);
+					}
+				}
+			}
+			if (previousPair != null && nextPair != null) {
+				float timeFromLastPair = this.tickCount - previousPair.getKey().intValue();
+				float timeBetweenPairs = nextPair.getKey().intValue() - previousPair.getKey().intValue();
+				float sizeDifference = nextPair.getValue().floatValue() - previousPair.getValue().floatValue();
+				//radius = ((Float) previousPair.getValue()).floatValue() + sizeDifference * (timeFromLastPair / timeBetweenPairs);
+				getEntityData().set(RADIUS, (timeFromLastPair / timeBetweenPairs) * sizeDifference + (previousPair.getValue().floatValue()));
+			}
+			
+			//handle entity collisions
+			List<LivingEntity> entityList = level.getEntitiesOfClass(LivingEntity.class, this.getBoundingBox());
+			for (LivingEntity entity : entityList) {
+				if (Math.sqrt(Math.pow(entity.getX() - this.getX(), 2) + Math.pow(entity.getZ() - this.getZ(), 2)) <= getEntityData().get(RADIUS)) { //better precision (makes real hitbox round)
+					for (EffectInstance effect : isEntityException(entity) ? secondaryEffects : primaryEffects) { //decide which effects to apply
+						entity.addEffect(effect);
+					}
+				}
+			}
+		}
+	}
+	
+	private boolean isEntityException(LivingEntity entity) {
+		if (isOwnerException && entity == getEntityFromUUID(ownerUUID)) return true;
+		for (UUID currentUUID : exceptionUUIDs) {
+			if (entity.getClass() == getEntityFromUUID(currentUUID).getClass()) return true;
```

Comment:
getEntityFromUUID(currentUUID).getClass() will crash if there won't be any entity with such id (in case it was unloaded, etc)

Reply:
and why do u use UUID as filter, when you later check for class? If you can filter by type, then use Class<?. everywhere. Otherwise use entity.getUUID().equal(...)